### ISSUE TITLE
Scrolling SDK (virtual-scroll) - View isn't updated when list is updated (items added/removed)

### ISSUE THREAD

**Issue by Gil-Epshtain** [angular/material2#14635]
#### What is the expected behavior?
- If a list is bind to `<cdk-virtual-scroll-viewport>`, and later items will be added/ removed from this list, the `<cdk-virtual-scroll-viewport>` updated as well.

#### What is the current behavior?
Changing the list will not updated the view (`<cdk-virtual-scroll-viewport>`)

#### What are the steps to reproduce?
StackBlitz starter: https://stackblitz.com/edit/angular-hsr4tz?file=app%2Fcdk-virtual-scroll-overview-example.html<br/>
In this demo you can see that after adding items to the list, the view isn't updated.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Angular 7.1.4
Material 7.2.0
Browser: Chrome

#### Is there anything else we should know?
I have a feeling that this is a wrong use issue. However I could not find anywhere in the docs, how should I add or remove items.
See docs at: https://material.angular.io/cdk/scrolling/overview

**Comment by klew582** [comment]
when will this fix be released?

**Comment by elgs** [comment]
Try this:
```
add() {
    this.items = [...this.items];
    this.items.push('NEW - ' + this.items.length);
}
```

https://stackblitz.com/edit/angular-hsr4tz-mz26wq

**Comment by Gil-Epshtain** [comment]
@elgs - I like your answer, you can even make it one-liner with:
`add() {
    this.items = [...this.items, ('NEW - ' + this.items.length)];
}`

However, you need to remember that "virtual-scroll" directive purpose  is to make an extremely long (thousands items or even more) list to be more smooth. But if each time you need to "Add" an item, you regenerate the whole list... then you lose the purpose of this directive, since re-creating the whole list each time you add an item, is very inefficient.

**Comment by elgs** [comment]
@Gil-Epshtain agreed!

@crisbeto any chance to get any update about this fix? Thanks.

**Comment by elgs** [comment]
@Gil-Epshtain from my experience, this workaround is actually not bad, I mean, performance wise. The selling point of the virtual scroll is to swap in/swap out rendered items from the backed full item list. With this workaround, the backed full item list got completely redone, but the rendered window (of 50, let's say) of items remains 50. It may no be idea, but this actually does not affect the selling point of the scroll view. My $.02.

**Comment by elgs** [comment]
> However, you need to remember that "virtual-scroll" directive purpose is to make an extremely long (thousands items or even more) list to be more smooth. 

This statement is true.

> But if each time you need to "Add" an item, you regenerate the whole list... then you lose the purpose of this directive, since re-creating the whole list each time you add an item, is very inefficient.

No, you missed the point how the scroll view works. The scroll view has two parts: 1) The backing full list of items (could be thousands of), 2) the small amount of rendered items in DOM (like 50 items). The point of the scroll view is to prevent the thousands of items from being poured in the DOM. Copying the references of each items from the old backed list to a new list is actually cheap (though not ideal), the expensive part is rendering DOMs in the browser. In this case, you only redo the backed list, but the number of rendered DOM remain unchanged, so you don't lose the purpose of this directive.

**Comment by Gil-Epshtain** [comment]
@elgs - I think the best way to test this is with a real world example - once I've implement this - I'll return with conclusions

**Comment by RenatOliveirra** [comment]
@Gil-Epshtain you can force a detect changes to update the virtual scroll from ChangeDetectionRef.detectChange(), this method force the view to detect the changes. That can update the virtual-scroll list;

     @Component({
          selector: 'app-component',
          templateUrl: './app-component.html',
          styleUrls: ['./app-component.scss'],
          changeDetection: ChangeDetectionStrategy.OnPush /* This is a strategy to detect changes by 
          demand,  it isn't necessary to your case;*/
     })

     items: AppComponentItems[] = [];

     constructor(changeDetectorRef: ChangeDetectorRef){}

     addItem(){
          items.push(newItem);
          changeDetectorRef.detectChange();
     }

**Comment by murtuzamacdev** [comment]
> @Gil-Epshtain you can force a detect changes to update the virtual scroll from ChangeDetectionRef.detectChange(), this method force the view to detect the changes. That can update the virtual-scroll list;
> 
> ```
>  @Component({
>       selector: 'app-component',
>       templateUrl: './app-component.html',
>       styleUrls: ['./app-component.scss'],
>       changeDetection: ChangeDetectionStrategy.OnPush /* This is a strategy to detect changes by 
>       demand,  it isn't necessary to your case;*/
>  })
> 
>  items: AppComponentItems[] = [];
> 
>  constructor(changeDetectorRef: ChangeDetectorRef){}
> 
>  addItem(){
>       items.push(newItem);
>       changeDetectorRef.detectChange();
>  }
> ```

Did not work for me.

**Comment by alexdabast** [comment]
> @Gil-Epshtain you can force a detect changes to update the virtual scroll from ChangeDetectionRef.detectChange(), this method force the view to detect the changes. That can update the virtual-scroll list;
>  @Component({
>       selector: 'app-component',
>       templateUrl: './app-component.html',
>       styleUrls: ['./app-component.scss'],
>       changeDetection: ChangeDetectionStrategy.OnPush /* This is a strategy to detect changes by 
>       demand,  it isn't necessary to your case;*/
>  })
> 
>  items: AppComponentItems[] = [];
> 
>  constructor(changeDetectorRef: ChangeDetectorRef){}
> 
>  addItem(){
>       items.push(newItem);
>       changeDetectorRef.detectChange();
>  }
> 
> 
> Did not work for me.

Same for me changeDetectorRef is not working

**Comment by literalpie** [comment]
I don't think this is a bug. Change detection is not intended to be "deep" checking each item in the array. It just checks the reference. Because you are using `push`, the reference does not change. You should use deconstructing like @elgs mentioned. You can use `trackBy` to prevent re-rendering the whole list

**Comment by alexdabast** [comment]
> I don't think this is a bug. Change detection is not intended to be "deep" checking each item in the array. It just checks the reference. Because you are using push, the reference does not change. You should use deconstructing like @elgs mentioned. You can use trackBy to prevent re-rendering the whole list

I'm wondering why a regular *ngFor can detect changes on a push ?

**Comment by moneychaudhary** [comment]
Help:: After assigning list as new object, it moves the scrolling position to the top and not retaining the scrolling position.

**Comment by MayuriRathod** [comment]
Is the issue fixed?
when will this be available for us?

**Comment by medTL** [comment]
this.items = [...this.items, ('NEW - ' + this.items.length)]; worked with me, but how to replace existing object with new one

**Comment by wernert** [comment]
now it is 8.2021 and this issue is still not be fixed ?
and i found no mention about this unexpected behavior in the docs.
I for sure have not the intention to make a copy of all the data on every update 
this makes the whole virtual scroll quite irreal.
Please what is the reason for this strategy, do i miss something  ?

**Comment by elgs** [comment]
@wernert there is a trade off. Change detection is expensive, even so for a list. I think you best bet would be to use the `[...someArray]`.

**Comment by samherrmann** [comment]
I just had a use case where the immutable array approach (`[...someArray]`) was not feasible due to frequent updates to a large array. Luckily [*cdkForOf](https://material.angular.io/cdk/scrolling/api#CdkVirtualForOf) also accepts the array as an Observable. That means you can mutate the array and the Observable successfully notifies `*cdkForOf` when it needs to update:

```html
<cdk-virtual-scroll-viewport>
  <div *cdkVirtualFor="let item of items$">{{ item }}</div>
</cdk-virtual-scroll-viewport>

<button (click)="add()">Add item</button>
``` 

```ts
  items$ = new BehaviorSubject(
    Array.from({ length: 100 }).map((_, i) => this.item(i))
  );

  add() {
    const items = this.items$.getValue();
    items.push(this.item(items.length));
    this.items$.next(items);
  }

  private item(index: number): string {
    return `Item #${index}`;
  }
```

[StackBlitz Example](https://stackblitz.com/edit/angular-hsr4tz-dxledf?file=app%2Fcdk-virtual-scroll-overview-example.html)

**Comment by ChaitanyaBabar** [comment]
Approach suggested by @samherrmann worked for me.
**Saved my hours of debugging.**

**Comment by sabindev** [comment]
> I just had a use case where the immutable array approach (`[...someArray]`) was not feasible due to frequent updates to the large array. Luckily [*cdkForOf](https://material.angular.io/cdk/scrolling/api#CdkVirtualForOf) also accepts the array as an Observable. That means you can mutate the array and the Observable successfully notifies `*cdkForOf` when it needs to update:
> 
> ```
> <cdk-virtual-scroll-viewport>
>   <div *cdkVirtualFor="let item of items$">{{ item }}</div>
> </cdk-virtual-scroll-viewport>
> 
> <button (click)="add()">Add item</button>
> ```
> 
> ```ts
>   items$ = new BehaviorSubject(
>     Array.from({ length: 100 }).map((_, i) => this.item(i))
>   );
> 
>   add() {
>     const items = this.items$.getValue();
>     items.push(this.item(items.length));
>     this.items$.next(items);
>   }
> 
>   private item(index: number): string {
>     return `Item #${index}`;
>   }
> ```
> 
> [StackBlitz Example](https://stackblitz.com/edit/angular-hsr4tz-dxledf?file=app%2Fcdk-virtual-scroll-overview-example.html)

u are a legend

**Comment by RonStrauss** [comment]
If none of the solutions here helped you, you might try reducing the `templateCacheSize`. I couldn't see updates but playing with the cache finally helped for me.

[View Recycling](https://material.angular.io/cdk/scrolling/overview#view-recycling)
